### PR TITLE
autotest: testing userspace interrupts

### PR DESCRIPTION
**Pull request by pthierry-ledger**
Userspace interrupts (IRQn hold and declared by userspace drivers) need to be dynamically tested in order to validate that the user interrupts subsystem of Sentry do work.
In order to stay as most generic as possible, a generic timer driver is used at autotest level in order to configure and wait for timer interrupts, with a period match. ST Basic timers are ST-generic IPs that exists in, AFAIK, all STM32 family. This means that the autotest backend, that need to include a timer driver, will not need multiple implementations on STM32 familly. On other manufacturer (NXP, etc.) timers are most of the timer one of the core IPs that are usually reused from SoC to SoC, at least for simple timers, avoiding multiplication of drivers in autotest, which is not made for this.

In order to ensure enough portability, the timer drivers should alias all its symbols and ensure a clean enough API so that there is no SoC specific input in autotest frontend and test suite.

In this PR, we use the TIM6 timer from u5a5, yet, depending on the SoC and manufacturer, any timer should be abstracted with an homogeneous API to allow TEST_IRQ in various SoCs.

- [x] Adding timer 5 to autotest device tree for u5
- [x] Adding generic st-timers driver that configure a general purpose timer in order to periodically emit an interrupt
- [x] Write all interrupt-related test suite, that configure a timer, enable and check interrupts and then deactivate interrupts
- [x] Validate testsuite on u5
- [x] Validate that the very same driver do work on other ST SoCs (F4, L4...) only using specific dts-configuration

NOTE: UAPI updated documentation and robotframework test are out of this PR for the sake of simplicity.

### CODE REVIEW

File: autotest/include/drivers/timer.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// SPDX-FileCopyrightText: 2023 Ledger SAS
```
Comment (fvalette-ledger) [inline autotest/include/drivers/timer.h:1]: Same as above, apply to all following occurrence

File: kernel/src/managers/device/device.c [hunk 2]
```diff
@@ -122,6 +122,8 @@ kstatus_t mgr_device_init(void)
             /* owner is not a task */
             owner = 0;
         } else {
+            #ifndef CONFIG_BUILD_TARGET_AUTOTEST
```
Comment (fvalette-ledger) [inline kernel/src/managers/device/device.c:125]: left aligned cpp clause

**Review by fvalette-ledger — APPROVED** [review]

File: autotest/include/arch/asm-cortex-m/io.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
+// SPDX-FileCopyrightText: 2023 Ledger SAS
```
[the remark this answers is not in the document]
Comment (pthierry-ledger) [reply autotest/include/arch/asm-cortex-m/io.h:1]: I really need to add the auto-spdx config :-p

File: dts/examples/stm32f429i_disc1_autotest.dts [hunk 4]
```diff
@@ -66,6 +66,19 @@
 	};
 };
 
+&timers5 {
```
Comment (iartemov-ledger) [inline dts/examples/stm32f429i_disc1_autotest.dts:69]: TIM5 seems to have the position 50 (not 49) in STM32F42xxx reference manual (it is 49 for STM32U5).

File: autotest/src/tests/test_irq.c [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    /* waiting 1200ms */
+    res = sys_wait_for_event(EVENT_TYPE_IRQ, 0);
+    copy_to_user(&tab[0], sizeof(exchange_event_t) + 4);
+    ASSERT_EQ(res, STATUS_OK);
+    IRQn = (uint32_t*)&((exchange_event_t*)tab)->data;
+    ASSERT_EQ(*IRQn, 49);
+    return;
+}
+
+static void test_irq_spawn_one_it(void)
```
Comment (iartemov-ledger) [inline autotest/src/tests/test_irq.c:40]: More logical if the TC goes first?

File: autotest/src/tests/test_irq.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+{
+    Status res;
+    uint8_t tab[128];
+    uint8_t count;
+    timer_enable_interrupt();
+    timer_set_periodic();
+    timer_enable();
+
+    for (count = 0; count < 5; ++count) {
+        /* reeanble interrupt line (nvic unmasked)*/
```
Comment (iartemov-ledger) [inline autotest/src/tests/test_irq.c:67]: re-enable ?

File: autotest/src/drivers/timer/stm32-basic-timer.c [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    return 0;
+}
+
+
+int timer_disable_interrupt(void)
+{
+    uint32_t reg = 0;
+    stm32_timer_desc_t const *desc = stm32_timer_get_desc();
+    /* disable interrupt update */
+    iowrite32(desc->base_addr + TIM_DIER_REG, reg);
```
Comment (iartemov-ledger) [inline autotest/src/drivers/timer/stm32-basic-timer.c:102]: No need to call `sys_irq_disable()` ?

File: autotest/src/tests/test_irq.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+static void test_irq_spawn_two_it(void)
+{
+    Status res;
+    uint8_t tab[128];
+    uint32_t *IRQn;
+    timer_enable_interrupt();
+    timer_enable();
+    /* waiting 1200ms */
+    res = sys_wait_for_event(EVENT_TYPE_IRQ, 0);
```
Comment (iartemov-ledger) [inline autotest/src/tests/test_irq.c:24]: Never need to call `sys_irq_acknowledge()` ?

File: dts/examples/stm32f429i_disc1_autotest.dts (hunk 4 above)
Comment (pthierry-ledger) [reply dts/examples/stm32f429i_disc1_autotest.dts:69]: yep, defined in official dtsi, not here :)

**Pull request merged**